src/include/random.h

```diff
@@ -133,105 +161,39 @@ void randomize_rng()
   detail::randomize_rng<EngineT>();
 }
 
-template <typename IntegerT = int,
-          typename DistributionT = std::uniform_int_distribution<IntegerT>,
+template <typename NumberT = int,
+          typename DistributionT = detail::default_distribution<NumberT>,
           typename EngineT = std::default_random_engine>
-IntegerT generate_random_number()
+NumberT generate_random_number()
 {
-  using limits = std::numeric_limits<IntegerT>;
-  return detail::generate_random_number<IntegerT, DistributionT, EngineT>
-          (limits::min(), limits::max());
+  return detail::generate_random_number<NumberT, DistributionT, EngineT>
+          (0, std::numeric_limits<NumberT>::max());
```

Comment:
was this change from `limits::min()` back to `0` here intentional? we discussed this case previously [here](https://github.com/ceph/ceph/pull/17972#pullrequestreview-67394174)

i understand your point that `min=0` makes the interface closer to the `rand()` function that it aims to replace, but i think that a more explicit interface would be less confusing and prone to bugs in the long run

so first, i'd argue for removing the `generate_random_number(max)` overloads entirely, and require the caller to pass `(0, max)` instead:
* it only costs 2 extra characters of typing, and there's no confusion at the call site about the range of output values
* it eliminates confusion about whether this single-argument overload is the same as `generate_random_number(min, max)` but with a default argument for `max=limits::max()` (which was my first guess)
* it avoids the case of invalid input, like what happens if you call `generate_random_number(-5)`?

i see this `generate_random_number()` overload as a useful way to avoid typing out all of the `std::numeric_limits<>` crap, but setting `min=0` seems arbitrary to me - if anything, it should use the full range of values available to the given type. that said, there may be alternative ways to save typing, like providing aliases for some common distributions, and making first-class overloads for `generate_random_number(dist)` that take them. then you could do something like this: `generate_random_number(positive_ints);`

regardless, i think we should add overloads that take a distribution, and implement the (min, max) versions in terms of that. i still don't like the way we're using `thread_local` for distributions that could just as easily be on the stack and stored in registers. (i'm not sure exactly what the standard says about thread_local storage, but there's bound to be some overhead there)